### ISSUE TITLE
Add arbitrary formatting of time values in Property Expansion

### ISSUE THREAD

**Issue by Ud71p** [mpv-player/mpv#4117]
This is a feature request to extend the Property Expansion mechanism to allow for arbitrary formatting of time values.

For instance:
`osd-msg3='${time-remaining}'`
results in:
00:45:21

I would like to be able to format the time-remaining in some other format than HH:MM:SS, to suit my needs better. For instance I don't care about seconds at all, and I rarely see anything longer than 99 minutes, so I would like the format MM. This way I can show time-remaining in 2 characters instead of 8. For instance I would like to write:
`osd-msg3='${time-remaining%MM}'`

mpv has already time-formatting, I found one in the documentation of --screenshot-template argument. So what is missing is just some kind of glue-code to allow for arbitrary formats in property expansion.

Even better, to make the formatting really arbitrary, mpv could allow piping the value to arbitrary command. Then I could write a simple shell script that formats the data the way I want, like so:
`osd-msg3='${time-remaining|bash ~/.mpv/my_time_format.sh}'`

For instance I always wanted to display current wall time during movie playback, plus the time-remaining in MM format if MM>1 otherwise SS format. That would be hard with just property expansion, but trivial if piping is added.

**Comment by wiiaboo** [comment]
Why not just use a lua script? Even if you're not familiar with the language it's very quick to learn and there's a few examples in the mpv repo and User-Scripts wiki page to copy code from.

**Comment by Ud71p** [comment]
For security and performance reasons I have compiled mpv without lua. And the whole beauty of piping is: then I can use any language I like.

**Comment by wiiaboo** [comment]
You're worried about security and you're asking for random shell interpretation in property expansion?

**Comment by Ud71p** [comment]
Yes. There is no insecurity introduced. The user already can execute shell scripts and mpv, so she gains no new powers by executing shell scripts via mpv (or vice versa, executing mpv from a shell script). What I personally find insecure is installing a new package (lua, and all of its dependencies) on my system - that's more bugs to exploit. I'm trying to keep my system extremely minimalistic for the sake of security. And the performance issue is big for me, it's a very old hardware and I can barely play movies - any extra CPU usage (for lua) would probably mean the movie would no longer play smooth.

**Comment by wm4** [comment]
Lua has no dependencies and is more secure than having a shell installed. If you want security, you should probably uninstall ffmpeg and anything that uses it, including mpv.

**Comment by wiiaboo** [comment]
There's no performance penalty of including lua. By default it's only used for the OSC (which is relatively performance-heavy but you can disable with --no-osc) and the ytdl_hook (which also needs youtube-dl installed and if you do have it installed you can disable with --no-ytdl). Lua script are the easiest way to extend mpv's capabilities that doesn't involve messing with C and adding fringe features.

**Comment by Ud71p** [comment]
It's very interesting and valid points you are raising, but my feature request is small: mpv already has time value formatting (see --screenshot-template), and it also has arbitrary command execution (see 'run' in List of Input Commands). So it's just some glue between property expansion and these features that is needed. Of course it is completely up to you devs to decide whether you want to include any of these 2 things, but for me writing ${time-remaining%MM} makes more sense than installing lua, learning it, and figuring out how to use it with mpv to achieve the same effect.

**Comment by butterw** [comment]
Any update on this ?

If providing the full --screenshot-template time formatting options isn't possible, what about providing a single simplified time string ?

We already have 00:01:21 "${playback-time}", the ms variant 00:01:21.250 "${playback-time/full}" and 81.250000 "${=playback-time}".

Maybe all that is missing for typical use is 01:21 "${playback-time/short}" ? Mpc-hc automatically switches to this shorter time string.

It would be needed for display of playback-time, duration and  time-remaining.

**Comment by butterw** [comment]
> Why not just use a lua script?
1) more complex and less user friendly than properties.
2) potential collisions with other scripts if you register seek event.
3) performance impact. If time is displayed permanently on screen, it will need to be updated though the script each frame.

**Comment by UnkwUsr** [comment]
+1 for new format `playback-time/short` (proposed in https://github.com/mpv-player/mpv/issues/4117#issuecomment-1669808338). What opinion of project maintainers on it? Any chance it will me merged if someone propose pull request?
